### PR TITLE
Open chests when wielding corals

### LINKED ISSUE #2386: Coral can't open chests.

### PR DESCRIPTION
**Pull request by SmallJoker**
Code simplification by combining the `on_place` functions

Fixes #2386

### CODE REVIEW

File: mods/default/nodes.lua [hunk 1]
```diff
@@ -1986,6 +1986,45 @@ minetest.register_node("default:sand_with_kelp", {
...
+	end
+
+	if node_under.name ~= "default:coral_skeleton" or
+			minetest.get_node(pos_above).name ~= "default:water_source" then
+		return itemstack
+	end
+
+	if minetest.is_protected(pos_under, player_name) or
+			minetest.is_protected(pos_above, player_name) then
+		minetest.log("action", player_name
```
Comment (SmallJoker) [inline mods/default/nodes.lua]: Same as `core.item_place_node`. Protection mods will do the chat message if they want to.

File: mods/default/nodes.lua [hunk 2]
```diff
@@ -1986,6 +1986,45 @@ minetest.register_node("default:sand_with_kelp", {
...
+			.. " tried to place " .. itemstack:get_name()
+			.. " at protected position "
+			.. minetest.pos_to_string(pos_under))
+		minetest.record_protection_violation(pos_under, player_name)
+		return itemstack
+	end
+
+	minetest.swap_node(pos_under, {name = itemstack:get_name()})
```
Comment (paramat) [inline mods/default/nodes.lua]: I'm not sure we should use 'swap node', as coral skeleton might have metadata added to it for some reason.

File: mods/default/nodes.lua [hunk 3]
```diff
@@ -1986,6 +1986,45 @@ minetest.register_node("default:sand_with_kelp", {
...
+
+	local player_name = placer:get_player_name()
+	local pos_under = pointed_thing.under
+	local pos_above = pointed_thing.above
+	local node_under = minetest.get_node(pos_under)
+	local def_under = minetest.registered_nodes[node_under.name]
+
+	if def_under and def_under.on_rightclick then
+		return def_under.on_rightclick(pos_under, node_under.name,
+				placer, itemstack, pointed_thing) or itemstack
```
Comment (paramat) [inline mods/default/nodes.lua]: The condition ensures that `def_under.on_rightclick` is not nil, so why use `or itemstack`? That is never needed?

**Comment by paramat** [comment]
+1 once line comments are addressed. Didn't test, i trust your testing.

File: mods/default/nodes.lua (hunk 3 above)
Comment (DS-Minetest) [reply mods/default/nodes.lua]: [`core.item_place` does this, too.](https://github.com/minetest/minetest/blob/5560ec8178053b27adac6f277f539975d46267b8/builtin/game/item.lua#L424) Even though it is not implicitly allowed, there might be some mods which simply don't return the itemstack in `on_rightclick`. And normally there won't be any issues with that.

File: mods/default/nodes.lua [hunk 4]
```diff
@@ -1986,6 +1986,45 @@ minetest.register_node("default:sand_with_kelp", {
...
+		return itemstack
+	end
+
+	local player_name = placer:get_player_name()
+	local pos_under = pointed_thing.under
+	local pos_above = pointed_thing.above
+	local node_under = minetest.get_node(pos_under)
+	local def_under = minetest.registered_nodes[node_under.name]
+
+	if def_under and def_under.on_rightclick then
```
Comment (DS-Minetest) [inline mods/default/nodes.lua]:
```suggestion
	if def_under and def_under.on_rightclick and not placer:get_player_control().sneak  then
```
You should never be able to trigger an `on_rightclick` callback with sneak.

File: mods/default/nodes.lua (hunk 3 above)
Comment (paramat) [reply mods/default/nodes.lua]: Ah i see, my mistake. Is fine then.

File: mods/default/nodes.lua (hunk 2 above)
Comment (SmallJoker) [reply mods/default/nodes.lua]: If mods inject arbitrary metadata into nodes it's their job to clean up their garbage after swapping. Swap node was made for situations like this, where metadata is either not present or has to be contained in the destination node.
You probably argued the same way with the moss growth, which would be far more important to use swap_node. Default does not need to spend more time on cleaning up garbage of other mods.

Comment (paramat) [reply mods/default/nodes.lua]: Thanks.

File: mods/default/nodes.lua (hunk 4 above)
Comment (paramat) [reply mods/default/nodes.lua]: Done.

**Comment by paramat** [comment]
:+1:

**Pull request merged**